salt/_modules/solutions.py

```diff
@@ -0,0 +1,115 @@
+'''
+Various utilities to handle solution.
+'''
+
+import re
+import logging
+import time
+import os
+import yaml
+
+import salt.utils.json
+from salt.exceptions import CommandExecutionError
+
+HAS_LIBS = True
+SOLUTIONS_CONFIG_MAP = 'solutions'
+SOLUTIONS_CONFIG_MAP_NAMESPACE = 'default'
+SOLUTIONS_CONFIG_FILE = '/etc/metalk8s/solutions.yaml'
+try:
+    import kubernetes.client
+    from kubernetes.client.rest import ApiException
+    from pprint import pprint
+except ImportError:
+    HAS_LIBS = False
+
+log = logging.getLogger(__name__)
+
+
+__virtualname__ = 'solutions'
+
+
+def __virtual__():
+    if HAS_LIBS:
+        return __virtualname__
+    else:
+        return False, 'python kubernetes library not found'
+
+def get_solutions_from_configmap(
+        context="kubernetes-admin@kubernetes",
+        kubeconfig="/etc/kubernetes/admin.conf"
+        ):
+    """Get list of imported solution from solutions config map.."""
+    cfg = __salt__['metalk8s_kubernetes.setup_conn'](
+        context=context,
+        kubeconfig=kubeconfig
+    )
+    api_instance = kubernetes.client.CoreV1Api()
+    try: 
+        api_response = api_instance.read_namespaced_config_map(
+            SOLUTIONS_CONFIG_MAP,
+            SOLUTIONS_CONFIG_MAP_NAMESPACE,
+            pretty=True, exact=True, export=True)
+        # The data dict is in form dict[solutionid.attribute] = value
+        # we want to convert it into dict[solutionid] = dict[attribute]=value
+        solutions = {}
+        if api_response.data:
+            for k, v in api_response.data.items():
+                sol_id, att = k.split('.')
+                sol_dict = solutions.setdefault(sol_id, {})
+                sol_dict[att] = v
+        return solutions
+    except ApiException as exc:
+        log.exception(
+            'Exception when calling '
+            'CoreV1Api->read_namespaced_config_map')
+        raise CommandExecutionError(exc)
+
+def get_solutions_form_configfile():
+    """Get list of solutions defined in solution.yaml."""
+    solutions = {}
+    if os.path.exists(SOLUTIONS_CONFIG_FILE):
+        with open(SOLUTIONS_CONFIG_FILE, 'r') as fd:
+            solutions = yaml.safe_load(fd)
+    else:
+        log.exception(
+            'Cannot find {}'.format(SOLUTIONS_CONFIG_FILE))
+        raise CommandExecutionError()
+    return solutions
+
+def get_unconfigured_solutions():
+    """
+    Return the difference between the configmap (configured) and the config 
+    file (required).
+    """
+    configmap_solutions = get_solutions_from_configmap()
+    config_file_solutions = get_solutions_form_configfile()
+    result = {k: config_file_solutions[k]
+              for k in config_file_solutions if k not in configmap_solutions}
+    print(result)
+    return result
+
+def set_configured(
```

Comment:
Not really like the naming here. It's a module for me it should be `add_solution` or something like that